cips/cip-28.md

```diff
@@ -0,0 +1,34 @@
+| cip | 28 |
+| - | - |
+| title | Transaction size limit |
+| description | Set limit for transaction size |
+| author | Josh Stein ([@jcstein](https://github.com/jcstein)), Nina Barbakadze ([@ninabarbakadze](https://github.com/ninabarbakadze)), Rootul Patel ([@rootulp](https://github.com/rootulp)) |
+| discussions-to | <https://forum.celestia.org/t/cip-limit-number-of-pfbs-and-non-pfbs-per-block-increase-transaction-size-limit/1843> |
+| status | Draft |
+| type | Standards Track |
+| category | Core |
+| created | 2024-10-16 |
+
+## Abstract
+
+This CIP proposes to set the limit for transaction size. The proposal is to set the transaction size limit to 2MiB. Setting the transaction size limit is consensus-breaking.
+
+## Specification
+
+1. Transaction size is limited to 2MiB by setting the versioned parameter `MaxTxSize` to 2097152, which is 2MiB in bytes. From version v3 and above, in `CheckTx`, `PrepareProposal`, and `ProcessProposal`, each transaction's size is checked against the `appconsts.MaxTxSize` threshold. This ensures that transactions over the limit are rejected or excluded at all stages, from initial submission to execution.
+
+## Rationale
+
+The rationale for this proposal is to set the transaction size limit to 2MiB, even with 8MiB blocks, to prevent issues with gossiping large transactions. Gossiping an 8MiB transaction without chunking could be detrimental to the network. This is a consensus-breaking change.
```

Comment:
```suggestion
The proposal to set the transaction size limit to 2MiB, even with blocks of 8MiB or larger, is primarily preventative. Gossiping transactions approaching 8MiB without chunking could be detrimental to the network. While this could be implemented as a non-consensus breaking change by modifying the default mempool `MaxTxBytes`, a consensus-breaking change was chosen for the following reasons:

1. Enforces a hard limit across all nodes, preventing discrepancies.
1. Provides stronger safeguards against potential network issues from large transactions.
1. Allows for controlled future experimentation with transaction sizes.

This approach prioritizes network-wide consistency and stability, but requires a coordinated network upgrade.
```